homeassistant/components/freebox/router.py

```diff
@@ -0,0 +1,381 @@
+        """Return the router MAC address."""
+        return self._mac
+
+    @property
+    def firmware_version(self) -> str:
+        """Return the router software version."""
+        return self._sw_v
+
+    @property
+    def device_info(self) -> Dict[str, any]:
+        """Return the device information."""
+        return {
+            "connections": {(CONNECTION_NETWORK_MAC, self.mac)},
+            "identifiers": {(DOMAIN, self.mac)},
+            "name": self.name,
+            "manufacturer": self.manufacturer,
+            "sw_version": self.firmware_version,
+        }
+
+    @property
+    def devices(self) -> Dict[str, any]:
+        """Return all devices."""
+        return self._devices
+
+    @property
+    def sensors(self) -> Dict[str, any]:
+        """Return all sensors."""
+        return {**self._temperature_sensors, **self._connection_sensors}
+
+    @property
+    def wifi(self) -> Wifi:
+        """Return the wifi."""
+        return self._api.wifi
+
+
+class FreeboxDevice:
+    """Representation of a Freebox device."""
+
+    def __init__(self, device: Dict[str, any]):
+        """Initialize a Freebox device."""
+        self._name = device["primary_name"].strip() or DEFAULT_DEVICE_NAME
+        self._mac = device["l2ident"]["id"]
+        self._manufacturer = device["vendor_name"]
+        self._icon = icon_for_freebox_device(device)
+
+        self.update(device)
+
+    def update(self, device: Dict[str, any]) -> None:
+        """Update the Freebox device."""
+        self._active = device["active"]
+        if device.get("attrs") is None:
+            # device
+            self._reachable = device["reachable"]
+            self._attrs = {
+                "reachable": self._reachable,
+                "last_time_reachable": datetime.fromtimestamp(
+                    device["last_time_reachable"]
+                ),
+                "last_time_activity": datetime.fromtimestamp(device["last_activity"]),
+            }
+        else:
+            # router
+            self._attrs = device["attrs"]
+
+    @property
+    def name(self) -> str:
+        """Return the name."""
+        return self._name
+
+    @property
+    def mac(self) -> str:
+        """Return the MAC address."""
+        return self._mac
+
+    @property
+    def manufacturer(self) -> str:
+        """Return the manufacturer."""
+        return self._manufacturer
+
+    @property
+    def icon(self) -> str:
+        """Return the icon."""
+        return self._icon
+
+    @property
+    def active(self) -> bool:
+        """Return true if the host sends traffic to the Freebox."""
+        return self._active
+
+    @property
+    def reachable(self) -> bool:
+        """Return true if the host can receive traffic from the Freebox."""
+        return self._reachable
+
+    @property
+    def state_attributes(self) -> Dict[str, any]:
+        """Return the attributes."""
+        return self._attrs
+
+
+class FreeboxSensor:
```

Comment:
What's the point of extracting the sensor interface to a separate class? It's using the same interface as the entity. Just put this in the entity class.

Reply:
Do you mean merging this `FreeboxSensor` and `sensor.FbxSensor` ? (as for devices)

Reply:
Yes.

The extra abstraction is not necessary as both classes are using the same interface, ie properties more or less. If there are any further internal attributes necessary they can just be internal attributes of the sensor entity.

Reply:
Ok, will do